pkg/server/server.go

```diff
@@ -96,19 +96,30 @@ func renderer(templatePath string) multitemplate.Renderer {
 	return r
 }
 
-func requireAuth(db *sqlx.DB, secret string) func(*gin.Context) {
+func getJobAuthSignature(ctx *gin.Context, db *sqlx.DB, secret string) (string, error) {
+	jobID := ctx.Param("id")
+	job, err := data.GetJob(jobID, db)
+	if err != nil {
+		log.Println(fmt.Errorf("getJobAuthSignature failed to getJob: %w", err))
+		ctx.AbortWithStatus(http.StatusInternalServerError)
+		return "", err
+	}
+
+	return job.AuthSignature(secret), nil
+}
+
+func requireTokenAuth(db *sqlx.DB, secret string, routeAuthCallback func(*gin.Context, *sqlx.DB, string) (string, error)) func(*gin.Context) {
 	return func(ctx *gin.Context) {
-		jobID := ctx.Param("id")
-		job, err := data.GetJob(jobID, db)
+		expected, err := routeAuthCallback(ctx, db, secret)
 		if err != nil {
-			log.Println(fmt.Errorf("requireAuth failed to getJob: %w", err))
+			log.Println(fmt.Errorf("requireAuth failed to get expected value: %w", err))
 			ctx.AbortWithStatus(http.StatusInternalServerError)
 			return
```

Comment:
Bug: This one is tricky. `getJobAuthSignature` will write a 500 response and return an error, which will also log and write a 500 here as well. Writing the header twice could result in some difficult to debug situations. Here's gin's warning about that: https://github.com/gin-gonic/gin/blob/master/response_writer.go#L63

Suggestion: instead of passing a func to `requireTokenAuth`, what about passing some value that would indicate that this is a job auth vs a user auth. Example:

```go
func requireAuth(db *sqlx.DB, secret, authType string) func(*gin.Context) {
	return func(ctx *gin.Context) {
		var expected string

		switch authType {
		case "job":
			jobID := ctx.Param("id")
			job, err := data.GetJob(jobID, db)
			if err != nil {
				log.Println(fmt.Errorf("getJobAuthSignature failed to getJob: %w", err))
				ctx.AbortWithStatus(http.StatusInternalServerError)
				return
			}
			expected = job.AuthSignature(secret)
		default:
			log.Println("requireTokenAuth failed, unexpected authType:", authType)
			ctx.AbortWithStatus(http.StatusInternalServerError)
			return
		}

		token := ctx.Query("token")

		// This is the same if it is a job or a user
		if token != expected {
			ctx.AbortWithStatus(403)
			return
		}
	}
}
```

Then when we're ready to implement the user auth flow, we just add another `authType` and switch case.

Alternatively, could just enforce that whatever function is passed in for `routeAuthCallback`, will not write anything back to the client.

Doubly Alternatively, could have two separate auth functions, one specifically for job, and another for user. 🤷‍♂️

In my opinion, passing a `ResponseWriter`, or in this case the `gin.Context`, to a helper function can end up with weird side effects and odd bugs. Restricting the ability to write back to the client within the immediate scope of the handler, or middleware in this case, by not passing the `ResponseWriter` to any helper functions, helps prevent those types of bugs. ❤️